pkg/reconciler/managed/reconciler.go

```diff
@@ -1096,28 +1115,123 @@ func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reco
 	return reconcile.Result{RequeueAfter: r.pollInterval}, errors.Wrap(r.client.Status().Update(ctx, managed), errUpdateManagedStatus)
 }
 
+// isManagementPolicySupported returns true if the management policy is
+// in the list of supported management policies. Short-circuits to true if
+// management policies are disabled.
+func isManagementPolicySupported(enabled bool, policy xpv1.ManagementPolicy) bool {
+	if !enabled {
+		return true
+	}
+	policySet := sets.New[xpv1.ManagementAction](policy...)
+	for _, p := range supportedManagementPolicies {
+		if p.Equal(policySet) {
+			return true
+		}
+	}
+	return false
+}
+
+// ContainsManagementAction returns true if the management policy contains
+// the management action or ManagementActionAll.
+func ContainsManagementAction(managementPolicy xpv1.ManagementPolicy, action xpv1.ManagementAction) bool {
+	for _, a := range managementPolicy {
+		if a == action || a == xpv1.ManagementActionAll {
+			return true
+		}
+	}
+	return false
+}
+
+// ContainsOnlyManagementAction returns true if the management policy contains
+// only the management action.
+func ContainsOnlyManagementAction(managementPolicy xpv1.ManagementPolicy, action xpv1.ManagementAction) bool {
+	return len(managementPolicy) == 1 && managementPolicy[0] == action
+}
+
+// shouldUpdate returns true if the management policy is enabled, and
+// contains the Update action.
+func shouldUpdate(managementPolicyEnabled bool, policy xpv1.ManagementPolicy) bool {
+	return shouldDoAction(managementPolicyEnabled, policy, xpv1.ManagementActionUpdate)
+}
+
+// shouldLateInit returns true if the management policy is enabled, and
+// contains the LateInitialize action.
+func shouldLateInit(managementPolicyEnabled bool, policy xpv1.ManagementPolicy) bool {
+	return shouldDoAction(managementPolicyEnabled, policy, xpv1.ManagementActionLateInitialize)
+}
+
+// shouldCreate returns true if the management policy is enabled, and
+// contains the Create action.
+func shouldCreate(managementPolicyEnabled bool, policy xpv1.ManagementPolicy) bool {
+	return shouldDoAction(managementPolicyEnabled, policy, xpv1.ManagementActionCreate)
+}
+
+// shouldDoAction returns true if the management policy is enabled, and
+// contains the supplied action.
+func shouldDoAction(managementPolicyEnabled bool, policy xpv1.ManagementPolicy, action xpv1.ManagementAction) bool {
+	if !managementPolicyEnabled {
+		return true
+	}
+	return ContainsManagementAction(policy, action)
+}
+
+// isManagementPolicyObserveOnly returns true if the management policy is
+// enabled, and contains only the Observe action.
+func isManagementPolicyObserveOnly(managementPolicyEnabled bool, policy xpv1.ManagementPolicy) bool {
+	if !managementPolicyEnabled {
+		return false
+	}
+	return ContainsOnlyManagementAction(policy, xpv1.ManagementActionObserve)
+}
```

Comment:
A lot of these functions take a policy and whether management policies are enabled as arguments. Would it make sense to introduce a `ManagementPolicyChecker` type? Might result in more succinct calls, and less refactoring when `managementPolicyEnabled` goes away (i.e. when its GA).

I'm thinking something like:

```go
type ManagementPolicyChecker interface {
	ShouldUpdate() bool
	ShouldLateInit() bool
	IsObserveOnly() bool  // Could this be phrased in some kind of ShouldFoo way?
}
```

You could then have one implementation that had both enabled and policy, e.g.:

```go
// This is a very dumb name.
type PolicyBasedManagementPolicyChecker struct {
	policy xpv1.ManagementPolicy
	enabled bool
}
```

Or one that embeds just `xpv1.ManagementPolicy` and a different implementation that is used whenever management policies aren't actually enabled.

Reply:
I like this. 

The interface could be extended with `IsNonDefault` and `IsSupported` as well.

Reply:
@lsviben Do you think you'll add `IsNonDefault` and `IsSupported` to this as well? I like @turkenh's suggestion - it would be nice for this type to be a one-stop-shop for all things management-policy-checking.

If we wanted to get reaaaally fancy it would be pretty clean to have a type that essentially loaded up all the context we need to make branching decisions (e.g management policies, deletion policy, paused annotation, etc etc). Something like:

```go
policy, err := NewManagementPolicyResolver(managed)
if err != nil {
    // NewManagementPolicyResolver returns an error if you're doing anything "wrong",
    //i.e. if you're using an unsupported management policy, or if you're trying to use
    // management policies but the feature flag isn't enabled.
}

type ManagementPolicyResolver interface {
    ShouldPause() bool
    ShouldOnlyObserve() bool

    ShouldCreate() bool
    ShouldDelete() bool  // Equivalent to !shouldOrphan
    ShouldLateInit() bool
    ShouldUpdate() bool
}
```

All the `ShouldFoo` functions don't feel quite as clean as the nice `Action` and `OnlyAction` methods you have that take a policy as arguments, but I think I still prefer them because they let us push all of the context needed to make "should I do this thing" checks (not only management policies) into the policy resolver.

Reply:
sure, i can add something like that, also like the `Resolver` name. Also like how it hides the checks for supported and the sanity check

Reply:
Added something similar:
```go
// ManagementPoliciesChecker is used to perform checks on management policies
// to determine specific actions are allowed, or if they are the only allowed
// action.
type ManagementPoliciesChecker interface {
	// IsSupported returns true if the management policy is supported.
	IsSupported() bool
	// IsDisabledNonDefault returns true if the management policy is set to a non-default
	// value, and the management policy feature is disabled. Used as a sanity
	// check.
	IsDisabledNonDefault() bool
	// IsPaused returns true if the resource is paused based
	// on the management policy.
	IsPaused() bool

	// ShouldOnlyObserve returns true if only the Observe action is allowed.
	ShouldOnlyObserve() bool
	// ShouldCreate returns true if the Create action is allowed.
	ShouldCreate() bool
	// ShouldLateInitialize returns true if the LateInitialize action is
	// allowed.
	ShouldLateInitialize() bool
	// ShouldUpdate returns true if the Update action is allowed.
	ShouldUpdate() bool
	// ShouldDelete returns true if the Delete action is allowed.
	ShouldDelete() bool
}
```

Added the IsPaused to it, but in the end this IsPaused just checks if its paused based on the management policies, and I left the meta.IsPaused to check paused condition based on annotations. Felt wrong to mix annotations in the ManagementPolicies checking tool, so i left them separate. 

Checking IsDisabledNonDefault and IsSupported through calls instead of an error in `NewManagement...` due to:
- if we want to have IsPaused there, we need to initialize it before it, so to avoid changing the order of checks from `IsPaused->IsDisabledNonDefault->IsSupported` to `IsDisabledNonDefault->IsSupported->IsPaused`
- simpler for testing

Just noticed that I didnt change naming, ill check that.

Reply:
> Felt wrong to mix annotations in the ManagementPolicies checking tool, so i left them separate.

FWIW despite using the name "management policy checker" in my examples I was thinking more of an "all the policies" checker as a single place to calculate the policy based on all the policy inputs, which right now I think are:

* Paused annotation
* Deletion policy
* Management policies

Though to be fair the first two will go away eventually at which point this would purely be checking management policies. I see your point WRT ordering of the paused/disabled/supported calls though. Don't feel super strongly on this one.

Another thought: Could we merge `IsDisabledNonDefault` and `IsSupported` into a `Validate() error` method that returns an error indicating what's wrong with the policy config (i.e. its set despite the flag being enabled, or its set to an invalid value)?